test/goHome.test.ts

```diff
@@ -0,0 +1,108 @@
+import { chromium, Page, Browser, BrowserContext, Cookie } from "playwright"
+import { createCookieIfDoesntExist } from "../src/common/util"
+import { hash } from "../src/node/util"
+
+describe("login", () => {
+  let browser: Browser
+  let page: Page
+  let context: BrowserContext
+
+  beforeAll(async (done) => {
+    browser = await chromium.launch()
+    // Create a new context with the saved storage state
+    const storageState = JSON.parse(process.env.STORAGE || "")
+
+    //
+    const cookieToStore = {
+      sameSite: "Lax" as const,
+      name: "key",
+      value: hash(process.env.PASSWORD || ""),
+      domain: "localhost",
+      path: "/",
+      expires: -1,
+      httpOnly: false,
+      secure: false,
+    }
+
+    // For some odd reason, the login method used in globalSetup.ts doesn't always work
+    // I don't know if it's on playwright clearing our cookies by accident
+    // or if it's our cookies disappearing.
+    // This means we need an additional check to make sure we're logged in.
+    // We do this by manually adding the cookie to the browser environment
+    // if it's not there at the time the test starts
+    const cookies: Cookie[] = storageState.cookies || []
+    // If the cookie exists in cookies then
+    // this will return the cookies with no changes
+    // otherwise if it doesn't exist, it will create it
+    // hence the name maybeUpdatedCookies
+    const maybeUpdatedCookies = createCookieIfDoesntExist(cookies, cookieToStore)
+    console.log("here are the cookies", maybeUpdatedCookies)
+
+    context = await browser.newContext({
+      storageState: { cookies: maybeUpdatedCookies },
+      recordVideo: { dir: "./test/videos/" },
+    })
+    done()
+  })
+
+  afterAll(async (done) => {
+    // Remove password from local storage
+    await context.clearCookies()
+
+    await browser.close()
+    await context.close()
+    done()
+  })
+
+  beforeEach(async (done) => {
+    page = await context.newPage()
+    done()
+  })
+
+  // NOTE: this test will fail if you do not run code-server with --home $CODE_SERVER_ADDRESS/healthz
+  it("should see a 'Go Home' button in the Application Menu that goes to /healthz", async (done) => {
+    // Ideally, this test should pass and finish before the timeout set in the Jest config
+    // However, if it doesn't, we don't want a memory leak so we set this backup timeout
+    // Otherwise Jest may throw this error
+    // "Jest did not exit one second after the test run has completed.
```

Comment:
Jest's error is correct. By setting this backup timeout, we're hiding that the test failed as the go home url wasn't requested.

This should be rewritten to use purely promises instead of the done callback. I'm guessing we can't combine them and so jest gets confused when you're returning a promise but also calling done later.

Reply:
> By setting this backup timeout, we're hiding that the test failed as the go home url wasn't requested.

Ahhh - now I see what you're saying. Oversight on my part. 

It's failing because we wait indefinitely until the request to the go home url has been made. This causes the memory leak because we never finish. I think a way forward to solve this is to add an expect call in the `backupTimeout` before `done` so we know that we've waited long enough for the request to happen to the go home url but that it was unsuccessful.

Reply:
Correction: I think you're right about Jest being confused using both promises and the done callback. I don't see the two used together [in the docs](https://jestjs.io/docs/en/asynchronous#callbacks) and I'm still getting the same leak error. I'll rewrite to use promises as suggested.